lib/ABI/FunctionType.cpp

```diff
@@ -388,210 +417,447 @@ class ConversionHelper {
-        } else {
-          return std::nullopt;
-        }
+        // Otherwise return a `void *` pointer as the return value type.
+        static constexpr auto VoidKind = model::PrimitiveTypeKind::Void;
+        auto Result = ensurePrimitiveTypeIsAvailable(VoidKind,
+                                                     0,
+                                                     Binary,
+                                                     TypeDependencies);
+        using model::Qualifier;
+        Result.Qualifiers.emplace_back(Qualifier::createPointer(Architecture));
+        return Result;
+      }
+    }
+
+    // TODO: there are some cases where preserving the total structure of
+    // the return value can be benefitial, this needs more work.
+    // For now, just determine the total size of the return value and if it
+    // fits into a primitive value - emit it, otherwise show those values as
+    // a struct (/note: positional ABIs need a whole lot of extra love in this
+    // regard).
+    llvm::SmallVector<std::optional<model::QualifiedType>, 4> ResultingTypes;
+    bool MustUseTheNextOne = false;
+    auto AllowedRange = llvm::enumerate(llvm::reverse(Allowed));
+    for (auto Pair : AllowedRange) {
+      size_t Index = Allowed.size() - Pair.index() - 1;
+      model::Register::Values Register = Pair.value();
+      auto UsedIterator = Used.find(Register);
+      if (UsedIterator != Used.end())
+        ResultingTypes.emplace_back(UsedIterator->Type);
+      else if (MustUseTheNextOne)
+        ResultingTypes.emplace_back(std::nullopt);
+
+      MustUseTheNextOne = MustUseTheNextOne || (UsedIterator != Used.end());
+    }
+
+    revng_assert(ResultingTypes.size() != 0);
+    if (ResultingTypes.size() == 1) {
+      // Short-circuit for a single register return value.
+      auto Kind = model::Register::primitiveKind(Allowed.at(0));
+      if (ResultingTypes.front() != std::nullopt) {
+        return ensureTypeIsAvailable(*ResultingTypes.front(),
+                                     Allowed.at(0),
+                                     Binary,
+                                     TypeDependencies);
+      } else {
+        // No information about the type of this register, use the default.
+        return ensureTypeIsAvailable(model::QualifiedType{},
+                                     Allowed.at(0),
+                                     Binary,
+                                     TypeDependencies);
+      }
+    } else if (ResultingTypes.size() == 2) {
+      // Short-circuit for a double register return value.
+      if (Allowed.size() < 2) {
+        // The ABI doesn't support double register return values.
+        // There's a good chance the ABI was not detected correctly.
+        return std::nullopt;
       }
+
+      auto UpperKind = model::Register::primitiveKind(Allowed.at(0));
+      auto LowerKind = model::Register::primitiveKind(Allowed.at(1));
+      if (UpperKind != LowerKind)
+        return std::nullopt;
```

Comment:
What do you mean? This is just a really little safety check for some unorthodox cases which could only result from a `RawFunctionType` being manually fiddled with.

In this case if, for example, there's a return value in `rax` marked as `PointerOrNumber` and one with `rdx` marked as `Generic`, I'd rather not touch those and abort the conversion instead - since the function at hand is non-standard.

Reply:
Ah, wait, no! I'm stupid and miss-understood my own code. Sorry about that. Scratch my last comment.

This is a check ensuring that if there are exactly two return value registers, they are no inherently of a different type, e.g. they are either both generic or floating point.

Reply:
OK, but this is basically a check on a property of an ABI description. I suggest we push this assumption further and verify it holds for all ABI descriptions, so that we don't have to verify it here.

Reply:
Actually, this is more of an artifact of the fact that we don't handle the floating point stuff at all. I imagine those would get filtered out separately before this "two register return value" branch is even entered on - since having 1 gpr and 1 vector register is probably a sign of a struct containing multiple members being returned.

I suggest we leave this check as is for now without wasting effort on trying to encode/verify this at the ABI level, and when the floating point is properly supported we get this check done implicitly (by selecting the right branch).

Reply:
I think with floating point stuff we'll have a separate set of return registers that we'll need to reason about separately?

But in any case, let's not leave around code "just in case", it makes the codebase confusing and then we have FUD when we ask ourselves if we can drop it. This happens a lot.
Let's define a scope, let's not crash in out-of-scope situations (i.e., soft fail) but let's not try to support things that are out of scope. Floating point are out of scope as of now.

Reply:
I guess there's an argument to leave this check here at least for the reason of "at the point when you try to merge two registers into a single type, make sure there's no indication that they are not compatible". But yeah, I'm ok with dropping this check for now.

Reply:
Up to you then

Reply:
I opted out to remove this check for now since it wasn't playing nicely with the new design where `RegisterStateDeductions` takes care of all the dirty work.

We'll still have to revisit this when we introduce a full-fledged vector register support.